lib/alchemy/locale.rb

```diff
@@ -31,11 +28,18 @@ def locale_change_needed?
       params[:locale].present? || session[:alchemy_locale].blank?
     end
 
-    # Try to get the locale from +params[:locale]+.
-    def locale_from_params
-      return if params[:locale].blank?
-      if ::I18n.available_locales.include?(params[:locale].to_sym)
-        params[:locale]
+    # Returns either the most preferred locale that is within the list of available locales or nil
+    #
+    # The availability of the locales is checked in the exact order of either
+    #
+    #  * the passed parameter: +params[:locale]+
+    #  * the user's locale
+    #  * the locale of the browser
+    #
+    def available_locale
+      locale_chain = [params[:locale], locale_from_user, locale_from_browser].compact
+      locale_chain.each do |locale|
```

Comment:
You want `detect` here. Because you want to return the first from `locale_chain` that's in available locales right?